### PR TITLE
Add context information into input runes

### PR DESCRIPTION
**Pull request by benoitkugler**
This PR adresses one of the remaining items from #1, by supporting start index and length so that the shaper can access surrounding context.

### CODE REVIEW

File: input.go [hunk 1]
```diff
@@ -8,7 +8,16 @@ import (
 
 type Input interface {
 	// Text returns the characters to be shaped.
-	Text() []rune
+	//
+	// `itemOffset` is the position of the first character from `text` that will be shaped, and
+	// `itemLength` is the number of character to shape (-1 means the end of the slice).
```
Comment (stuartmscott) [inline input.go]: nit: drop item prefix

File: input.go [hunk 2]
```diff
@@ -8,7 +8,16 @@ import (
...
+	//
+	// When shaping part of a larger text (e.g. a run of text from a paragraph),
```
Comment (stuartmscott) [inline input.go]: Probably just a line, not the whole paragraph as the text could be wrapped into multiple lines without explicit newline characters, and we wouldn't want the shaper to mistakenly shape the character at the start of a run using context of a character at the end of the previous run thinking they are adjacent when in fact they are on different lines/sides of the page.

File: input.go (hunk 1 above)
Comment (whereswaldon) [inline input.go]: My only question is whether using `-1` is really preferable to `len(input)` to specify the end of the list. It doesn't seem significantly more ergonomic to me, but I also haven't written much code that interacts with text shapers. I don't have a strong opinion about this though. I definitely don't mind if it stays this way.

**Review by whereswaldon — APPROVED** [review]
I think this makes sense. I have one thought about the API, but it isn't a very important one. I'd be happy for this to be merged in its current state.

File: input.go (hunk 1 above)
Comment (benoitkugler) [reply input.go]: Notice that with the convention of the PR, the second `int` is the _length_ of the segment, not the end index. 
Thus, if you want to shape text[2:end_of_the_slice], you would have to pass len(text) - 2 as `itemLength`, which is a bit less ergonomic to just pass `-1`.

We could also flip the convention to provide `(startIndex, endIndex)` if it is more natural...

Comment (whereswaldon) [reply input.go]: Ah, forgive me. I apparently cannot read. :D I'm fine with -1 as a convention then. It is totally more ergonomic. I would also be fine with the start and end convention, but I see no pressing need for it.

**Review by stuartmscott — CHANGES_REQUESTED** [review]
Thanks for putting this together @benoitkugler!

File: input.go (hunk 1 above)
Comment (stuartmscott) [reply input.go]: Mild preference of start, end as they can be used directly as `for i := start; i < end; i++` loops, without extra `if length == -1` checks

Comment (benoitkugler) [reply input.go]: Done

File: input.go (hunk 2 above)
Comment (benoitkugler) [reply input.go]: Good point, thanks ! Updated documentation.

File: input.go (hunk 1 above)
Comment (andydotxyz) [reply input.go]: Agreed - it seems to be more idiomatic of go as slices etc take that approach

**Review by stuartmscott — APPROVED** [review]
Good work, thanks!

**Comment by whereswaldon** [comment]
We have two approvals and no comments left to address. :tada:  I think we should merge in 24 hours if there are no further requests for changes.

**Comment by stuartmscott** [comment]
Agreed.

@benoitkugler are the permissions setup correctly for you to merge?

**Comment by benoitkugler** [comment]
> Agreed.
> 
> @benoitkugler are the permissions setup correctly for you to merge?

Yes thank you !

**Pull request merged**